File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/Payment/ApplePayStrategy.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.Payment;

/// <summary>
///     A payment strategy for Apple Pay.
/// </summary>
public sealed class ApplePayStrategy : PaymentStrategyBase
{
    private readonly string _appleUsername;

    /// <summary>
    ///     Initializes a new instance of the <see cref="ApplePayStrategy" /> class.
    /// </summary>
    /// <param name="appleUsername">The Apple Pay username.</param>
    /// <exception cref="ArgumentException">Thrown when username is empty.</exception>
    public ApplePayStrategy(string appleUsername)
    {
        if (string.IsNullOrWhiteSpace(appleUsername))
            throw new ArgumentException("Apple username cannot be empty.", nameof(appleUsername));
        _appleUsername = appleUsername;
    }

    /// <inheritdoc />
    public override void Checkout(InventoryItem item, int quantity, TextWriter output)
    {
        var total = ComputeTotalAmount(item, quantity);
        output.WriteLine($"Checking out {total:C} using Apple Pay (User: {_appleUsername}).");
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/payment/ApplePayStrategy.java
package assignment2solution.domain.payment;

import assignment2solution.domain.inventory.InventoryItem;

import java.io.PrintWriter;
import java.io.Writer;
import java.text.NumberFormat;
import java.util.Locale;
import java.util.Objects;

/**
 * A payment strategy for Apple Pay.
 */
public final class ApplePayStrategy extends PaymentStrategyBase {
    private final String appleUsername;
    private final NumberFormat currencyFormatter = NumberFormat.getCurrencyInstance(Locale.US);

    /**
     * Initializes a new instance of the {@link ApplePayStrategy} class.
     */
    public ApplePayStrategy(String appleUsername) {
        if (appleUsername == null || appleUsername.trim().isEmpty()) {
            throw new IllegalArgumentException("Apple username cannot be empty.");
        }
        this.appleUsername = appleUsername;
    }

    @Override
    public void checkout(InventoryItem item, int quantity, Writer output) {
        Objects.requireNonNull(item, "item");
        Objects.requireNonNull(output, "output");
        var writer = output instanceof PrintWriter ? (PrintWriter) output : new PrintWriter(output, true);
        var total = computeTotalAmount(item, quantity);
        writer.println("Checking out " + currencyFormatter.format(total) + " using Apple Pay (User: "
            + appleUsername + ").");
    }
}
